### PR TITLE
[opengl] GLFW 3.3.2 is now included as a submodule

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #664 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: cmake/TaichiCore.cmake [hunk 1]
```diff
@@ -62,12 +62,15 @@ if (TI_WITH_OPENGL)
...
-    target_link_libraries(${LIBRARY_NAME} glfw)
+
+    set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "Build the GLFW example programs")
+    set(GLFW_BUILD_TESTS OFF CACHE BOOL "Build the GLFW test programs")
+    set(GLFW_BUILD_DOCS OFF CACHE BOOL "Build the GLFW documentation")
+    set(GLFW_INSTALL OFF CACHE BOOL "Generate installation target")
+
+    message("Building with GLFW")
+    add_subdirectory(external/glfw)
+    target_link_libraries(${LIBRARY_NAME} glfw3)
```
Comment (archibate) [inline cmake/TaichiCore.cmake:73]: Will this add the `taichi/external/glfw/libglfw.a` instead of the `/usr/lib/libglfw.so`?

**Comment by archibate** [comment]
Great! It's building properly now! Could you also include GLEW 2.1.0 as submodule? See https://github.com/nigels-com/glew#recent-snapshots (we want it's buildable snapshot)

**Comment by archibate** [comment]
Btw I found `with_cuda` return true on my machine. And it failed to test.

**Comment by yuanming-hu** [comment]
> Great! It's building properly now! Could you also include GLEW 2.1.0 as submodule? See https://github.com/nigels-com/glew#recent-snapshots (we want it's buildable snapshot)

Given GLEW is not making use of CMake and it need special compilation, it no longer makes sense to include that as a submodule. The only solution is to ask the users to download the compiled objects and headers manually, or use the *ultimately* solution (https://github.com/taichi-dev/taichi/issues/664#issuecomment-612676371) to remove the dependency on GLEW.

**Comment by archibate** [comment]
我们可以打包一份**可编译**的glew源代码包给用户。

**Comment by yuanming-hu** [comment]
> 我们可以打包一份**可编译**的glew源代码包给用户。

That is probably very hard according to https://github.com/nigels-com/glew#build

**Comment by archibate** [comment]
一点也不hard，我只是make就完成了。

**Comment by yuanming-hu** [comment]
> 一点也不hard，我只是make就完成了。

I'm not sure if it's tricker on Windows.

It seems that they actually support CMake (which I assumed is more portably than `make`), but hidden in the `build` folder: https://github.com/nigels-com/glew/tree/master/build/cmake could you try if that works?

**Comment by archibate** [comment]
> ...could you try if that works?

Works!

**Comment by archibate** [comment]
所以我们要`add_subdirectory(external/glew/build/cmake)`?

**Comment by yuanming-hu** [comment]
> 所以我们要`add_subdirectory(external/glew/build/cmake)`?

Exactly!

**Comment by yuanming-hu** [comment]
> 所以我们要`add_subdirectory(external/glew/build/cmake)`?

If your network is not great, I can do that for you.

**Comment by archibate** [comment]
> If your network is not great, I can do that for you.

Already done. Thanks!

File: cmake/TaichiCore.cmake (hunk 1 above)
Comment (yuanming-hu) [reply cmake/TaichiCore.cmake]: Good question. I just now updated according to https://www.glfw.org/docs/3.3/build_guide.html#build_link_cmake_source.

**Pull request merged**